Inject small delay for busy workers to improve requests distribution

ayufan opened the pull request:
**Describe the bug**

We at GitLab.com are production testing Puma with prepartions for switching to it.
However, we noticed a few oddness when running Puma vs Unicorn in mostly similar configurations.

An worker of Puma when running in Cluster mode has no knowledge about siblings and their utilisation. This results in some of the sockets to be accepted by sub-optimal workers, that are already processing requests. This does have a statistical significance as in my local testing, and GitLab.com testing in the above workload we see an increase of around 20-30% of durations on P60%. Which is a significant increase.

**Capacity of Puma**

Capacity of Puma web-server is defined by `workers * threads`. Each of these
form a slot that can accept a new request. It means that each slot can accept
a new request at random (effectively round-robin).

Now, what happens if we have 2 requests waiting to be processed and two workers,
and two threads:

1. Our total capacity is 4 (W2 * T2),
2. Each request can be assigned at random to any worker, even the worker that is
   processing currently request, as we do not control that,
3. It means that in ideal scenario if we have 2 requests, for the optimal performance
   they should be assigned to two separate workers,
4. Puma does not implement any mechanism for 3., rather it is first accepting, first wins.
   It does mean that it is plausible that the two requests will be assigned in sub-optimal
   way: to the single worker, but multiple threads.
5. If the two requests are being processed by the same worker, they do share CPU-time,
   it means that they processing time is increased by the noisy neighbor due to Ruby MRI GVL.

Interestingly, the same latency impact is present on `Sidekiq`, we just don't see it,
as we do not care about real-time aspect of background processing that much.

However, the scheduling changes can improve `Sidekiq` performance as well if we would
target `sidekiq` as well.

This is more described in detail here: https://gitlab.com/gitlab-com/gl-infra/infrastructure/issues/8334#note_247859173

**Our Configuration**

We do run Puma and Unicorn in the following configurations:

1. Puma runs in `W16/T2`, the `16` corresponds to 16 CPUs available, nodes do not process anything else,
2. Unicorn runs in `W30`, the `30` seems like some artificial factor taken to have some sort of node saturation,
3. The nodes are lightly utilised, peeking at around 45% of CPU usage,
4. We seem to process around 70-80rq/sec.

We did notice oddness in Puma scheduling, as it was sometimes hitting the workers that are already processing requests. This resulted in increase of latency and duration of requests processing, even though we had a ton of spare capacity on other workers that were simply idle.

The nodes are configured today like that, due to graceful restart/overwrite/shutdown of Unicorn, to have a spare capacity to handle as many as twice amount of regular `W30`. They are highly underutilized on average due to that, but this is something to change. 

**Data**

The exact data are presented here: https://gitlab.com/gitlab-com/gl-infra/infrastructure/issues/8334#note_247824814 with the description of what we are seeing.

**Expected behavior: Ideal-scheduling algorithm in multi-threaded scenario**

In ideal scenario we would always want to use our capacity efficiently:

1. Assign to free workers first, as they do offer the best latency,
1. Assign to workers with the least amount of other requests being processed, as they are the least busy,
1. Assign to workers that are close to finish the requests being processed, as they will have a free capacity in the future (this is hard to implement, as this becomes a quite complex heuristic to understand the request completion time).

Currently, Puma does nothing from that. For round-robin we pay around 20% of ~performance penalty due to sub-optimal request processing assignment.

It is expected that the closer we get to 100% CPU-usage the more expected is that the Puma-non-scheduling algorithm will not be a problem, due to node being saturated, and threads by balanced by the kernel.

### Workaround

We can insert very small delay for busy workers, by assuming that they are already busy doing other work, and even if we schedule requests on them it will not really make them process requests faster.

The value of `5ms` is tested to provide a compromise between throughput and a delay, as we don't really want to delay forever. There's a test that tries to test that value and provide a benchmark which shows when this delay makes a difference.

### Your checklist for this pull request

- [ ] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [ ] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` to all commit messages.
- [ ] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [ ] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed, including Rubocop.

olleolleolle commented on lib/puma/thread_pool.rb:
```diff
@@ -213,6 +217,25 @@ def wait_until_not_full
+          puts "Injecting latency >>> tick:#{tick} busy:#{busy} backlog:#{@todo.size}."
+
+          # this will be signaled once a request finishes,
+          # which can happen ealier than tick time
```
typo: ealier -> earlier

evanphx commented:
This along with the original Issue are well researched and explained, thanks so much for that!

Would you be wiling to test a suite of values for latency and tick so we can get a feel for how they influence the average response time?

Could you run your test against a request that only lasts 150ms so we can get a feel for if your patch causes a meaningful slowdown in it? If it does, one option would be to sample average time to complete a job and feed it back into the wait algorithm so that it only ever attempts to sleep for, say, 1% or less of the average request time.

ayufan commented:
I work on doing some tests on GitLab.com to validate it :)

On Wed, 27 Nov 2019 at 06:52, Evan Phoenix <notifications@github.com> wrote:

> This along with the original Issue are well researched and explained,
> thanks so much for that!
>
> Would you be wiling to test a suite of values for latency and tick so we
> can get a feel for how they influence the average response time?
>
> Could you run your test against a request that only lasts 150ms so we can
> get a feel for if your patch causes a meaningful slowdown in it? If it
> does, one option would be to sample average time to complete a job and feed
> it back into the wait algorithm so that it only ever attempts to sleep for,
> say, 1% or less of the average request time.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/puma/puma/pull/2079?email_source=notifications&email_token=AASOSQKQZARQOEKMHKNN6KLQVYDILA5CNFSM4JQGH2X2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFIME6Q#issuecomment-558940794>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASOSQJD3PYLGXBLGP2XWBDQVYDILANCNFSM4JQGH2XQ>
> .
>

ayufan commented on lib/puma/server.rb:
```diff
@@ -214,6 +214,13 @@ def run_lopez_mode(background=true)
+    def wait_for_threads_to_finish
+      @thread_pool.wait_for_threads_to_finish(
+        tick_time: ENV.fetch('PUMA_INJECT_LATENCY', '0.001').to_f,
+        max_wait_ticks: ENV.fetch('PUMA_INJECT_WAIT_TICKS', '3').to_i
```
Not really. I was yet testing at all behaviour of max ticks or even if it is needed or if it helps with anything. I was assuming that after some time it simply does not matter. Worker is already so busy, that we are likely hitting saturation at this point.

The best would be to have some heuristic based on the current workload to know estimate of worker being freed, but this becomes too complex to handle. So, being dumb, likely doing 1 or 2 ticks should suffice in such cases.

schneems replied:
Env.fetch has some perf penalties. Can we read it from the environment once and store somwhere in a variable?

```ruby
require 'benchmark/ips'


variable = 0.001

Benchmark.ips do |x|
  x.report("env fetch") { ENV.fetch("FOO", 0.001) }
  x.report("variable") { variable }
  x.compare!
end

# Warming up --------------------------------------
#            env fetch   247.018k i/100ms
#             variable   382.650k i/100ms
# Calculating -------------------------------------
#            env fetch      5.572M (± 3.3%) i/s -     27.913M in   5.015719s
#             variable     21.348M (± 3.7%) i/s -    106.759M in   5.008674s

# Comparison:
#             variable: 21348220.1 i/s
#            env fetch:  5571895.6 i/s - 3.83x  slower
```

ayufan commented:
Yes, this approach particularly idle workers, or semi-idle. I did not see the other one to have big benefits for workloads and our configuration. I think it is too late to make that decision. The other one just prevents workers to pick new connections if they are already fully utilised (it seems), so it might seem that the other one is beneficial as well.

wjordan commented:
> I did not see the other one to have big benefits for workloads and our configuration.

By 'the other one' did you mean specifically the code in #1646 or the different code in #1920? Note that the code in #1920 is incorrectly implemented as I mentioned, and would have no effect.

ayufan commented:
Let me check my notes @wjordan, but also I was testing a very specific condition :) 

Yes. I was testing the #1920: https://docs.google.com/spreadsheets/d/1y2YqrPPgZ-RtjKiCGplJ7YkwjMXkZx6vEPq7prFbUn4/edit#gid=0. It shows no effect on my tests.

wjordan commented:
I tested out this PR on the Sinatra-MySQL TechEmpower benchmark mentioned in https://github.com/puma/puma/pull/1646#issuecomment-420845786:
- `m4.10xlarge` EC2 instance with 40 vCPUs
- Ruby 2.5.7-p206
- Puma configured with 40 workers, 5 threads (1 thread for comparison)
- Test configured for 40 concurrent requests, 20 queries per request (`./tfb --test sinatra --concurrency-levels 40 --type query --query-levels 20`)

Results (request/sec):

1 thread: **3917.86**

5 threads:
- `master`: **3119.52**
- `ayufan-research:gitlab-latency` (#2079, this PR): **3402.19**
- `wjordan:balance_requests` (#1646): **3932.26**

ayufan commented:
Thanks @wjordan. Let me check what happens with your changes when I run against my testing suite :)

ayufan commented:
## Puma BASELINE without patches

```
➜  gitlab-rails ./run-ab-projects
100	1	8.757	87.572
200	2	9.589	95.886
300	3	10.418	104.176
400	4	11.556	115.563
500	5	13.994	139.937
600	6	15.488	154.883
700	7	18.259	182.589
800	8	20.339	203.389
➜  gitlab-rails ./run-ab-projects
100	1	8.891	88.907
200	2	9.325	93.254
300	3	10.471	104.706
400	4	12.004	120.038
500	5	13.212	132.125
600	6	15.558	155.583
700	7	18.037	180.374
800	8	20.100	201.004
➜  gitlab-rails ./run-ab-projects
100	1	8.967	89.672
200	2	9.489	94.889
300	3	10.759	107.591
400	4	11.720	117.197
500	5	13.355	133.551
600	6	15.439	154.391
700	7	18.122	181.223
800	8	20.236	202.364
```

## PUMA with PR 2079

> https://github.com/puma/puma/commit/30f2206bb9d31bc9527b697f6948f7963315df88.patch

```
curl https://github.com/puma/puma/commit/30f2206bb9d31bc9527b697f6948f7963315df88.patch | patch -p1 -d /opt/gitlab/embedded/lib/ruby/gems/2.6.0/gems/puma-4.3.1/
```

```
➜  gitlab-rails ./run-ab-projects
100	1	9.506	95.056
200	2	9.586	95.860
300	3	10.281	102.806
400	4	10.811	108.114
500	5	13.427	134.265
600	6	16.204	162.037
700	7	18.505	185.049
800	8	21.025	210.249
➜  gitlab-rails ./run-ab-projects
100	1	8.932	89.320
200	2	9.015	90.146
300	3	9.749	97.495
400	4	10.317	103.170
500	5	12.891	128.913
600	6	15.084	150.837
700	7	17.667	176.668
800	8	20.239	202.388
➜  gitlab-rails ./run-ab-projects
100	1	9.185	91.852
200	2	9.243	92.427
300	3	9.557	95.567
400	4	10.269	102.689
500	5	12.625	126.253
600	6	15.184	151.837
700	7	17.795	177.945
800	8	19.925	199.245
➜  gitlab-rails ./run-ab-projects
100	1	8.879	88.789
200	2	9.056	90.557
300	3	9.596	95.961
400	4	10.370	103.698
500	5	13.024	130.237
600	6	15.276	152.763
700	7	17.490	174.896
800	8	20.303	203.026
```

### Puma with PR 1646

> https://patch-diff.githubusercontent.com/raw/puma/puma/pull/1646.patch

```
# I needed to fix manually, as it is not compatible with Puma 4.3.1
curl https://patch-diff.githubusercontent.com/raw/puma/puma/pull/1646.patch | patch -p1 -d /opt/gitlab/embedded/lib/ruby/gems/2.6.0/gems/puma-4.3.1/
```

```
➜  gitlab-rails ./run-ab-projects
100	1	9.353	93.534
200	2	10.097	100.967
300	3	11.138	111.383
400	4	12.623	126.234
500	5	14.353	143.528
600	6	16.087	160.866
700	7	18.354	183.539
800	8	21.022	210.224
➜  gitlab-rails ./run-ab-projects
100	1	8.861	88.609
200	2	9.608	96.078
300	3	11.007	110.065
400	4	11.523	115.233
500	5	13.487	134.872
600	6	15.591	155.913
700	7	18.424	184.245
800	8	20.288	202.876
➜  gitlab-rails ./run-ab-projects
100	1	8.918	89.179
200	2	9.255	92.552
300	3	10.456	104.563
400	4	11.964	119.635
500	5	13.714	137.145
600	6	16.010	160.098
700	7	17.923	179.234
800	8	20.582	205.822
➜  gitlab-rails ./run-ab-projects
100	1	8.928	89.279
200	2	9.604	96.035
300	3	10.933	109.333
400	4	11.741	117.414
500	5	13.128	131.279
600	6	15.790	157.905
700	7	18.070	180.696
800	8	20.592	205.915
```

## `run-ab-projects`

```bash
#!/bin/bash

run_ab() {
    result=$(ab -q -n "$1" -c "$2" http://localhost:4080/api/v4/projects)
    time_taken=$(echo "$result" | grep "Time taken for tests:" | cut -d" " -f7)
    time_per_req=$(echo "$result" | grep "Time per request:" | grep "(mean)" | cut -d" " -f10)

    echo -e "$1\t$2\t$time_taken\t$time_per_req"
}

for i in 1 2 3 4 5 6 7 8; do
    run_ab $((i*100)) $i
    sleep 1
done
```

ayufan commented:
I also put the results in: https://docs.google.com/spreadsheets/d/1y2YqrPPgZ-RtjKiCGplJ7YkwjMXkZx6vEPq7prFbUn4/edit#gid=1614859337.

ayufan commented:
We are running the `cond.wait` patch (this PR) on `GitLab.com` and we see a performance penalty in our testing. This is a change to before running patch.

ayufan commented:
We are running Puma on `gitlab.com` on 25% of fleet. So, far it really seems that patch does make Puma to be equal to Unicorn in our workload.

We are discussing that in: https://gitlab.com/gitlab-com/gl-infra/infrastructure/issues/8334#note_277452635.

The results from running today will be posted tomorrow, but so far distributions look pretty much equal between Puma/Unicorn. It means that threading does not affect for us on the request processing.

ayufan commented:
@nateberkopec @olleolleolle @wjordan 

It seems that simple change has a big impact for us: https://gitlab.com/gitlab-com/gl-infra/infrastructure/issues/8334#note_279516488. Doing extensive tests on `GitLab.com` shows no noticeable difference between Unicorn (W30/W40 (depends on service)) and Puma (W16/T4).

All our servers have 16 CPUs assigned.

I will look more into understanding @wjordan performance measurements.

nateberkopec commented:
I'm caught up on your thread on Gitlab now, but it looks like you still have some measurement issues re: the patch's effectiveness, correct?

ayufan commented:
@nateberkopec 

I would say not really, just comparison between one node vs another gives wrong results for various reasons: like traffic distribution, or variability in performance between nodes. This was also the reason why we started running at 23% of fleet to reduce these variability. 

Anyway, we will do another set of disables to have bigger sample sets. Now, it seems that the impact is substantial, and makes all execution duration to be pretty consistent with Unicorn (at least for us).

ayufan commented:
As of today GitLab.com runs fully on Puma with this patch: https://gitlab.com/gitlab-com/gl-infra/production/issues/1684.

I will now spend some time on comparing and figuring out which solution to ask for upstreaming :)

Thanks for your help!

schneems commented on lib/puma/const.rb:
```diff
@@ -100,7 +100,7 @@ class UnsupportedOption < RuntimeError
   module Const
 
-    PUMA_VERSION = VERSION = "4.3.0".freeze
+    PUMA_VERSION = VERSION = "4.3.1".freeze
```
I think you may need to rebase with master to get rid of this change. In general we'll rev the version.

schneems commented on History.md:
I think you may need to rebase with master to get rid of this change.

schneems commented on lib/puma/const.rb:
```diff
@@ -118,6 +118,13 @@ module Const
+    # but very "greedy" client across 10 requests. This prevents a not
+    # well behaved client from monopolizing the thread forever.
+    MAX_FAST_INLINE = 10
+
```
Need rebase

schneems commented on lib/puma/server.rb:
```diff
@@ -479,7 +490,19 @@ def process_client(client, buffer)
+              check_for_more_data = false
+            end
+
+            unless client.reset(check_for_more_data)
```
Need rebase.

schneems commented on lib/puma/thread_pool.rb:
```diff
@@ -213,6 +217,22 @@ def wait_until_not_full
+        end
+      end
+    end
+
```
This is the main meat-and-potatoes. The wait with timeout makes sense to me. Why do we need the "tick" here? What's the logic/reason behind doing this multiple times?

nateberkopec replied:
@schneems https://gitlab.com/gitlab-com/gl-infra/infrastructure/issues/8334#note_247859173

ayufan commented:
@schneems 

Thanks for review, this is to showcase approach, not something yet to be merged in that form (WIP), even though we already use it :) It was done like that to give us an ability to easily tinker via ENVs without changing any code.

I will remove the aspects that are not relevant: `env.fetch`, `ticks`, etc. to keep it truly minimal change.

ayufan commented:
The failure of `out_of_bound_stream` is due to: https://github.com/puma/puma/pull/2218#discussion_r413982179.

The test assumes that all requests are processed in parallel, ie. multiple threads is processing data in parallel, which might not always be true:
- scheduler might delay acceptance of the socket, till it is free even if it is has more threads
- test does fail if run `max_threads=1` which creates the sequence of execution, as test works on a failing edge
- the `out_of_bound` seems to hold a mutex which might result in the same contention problem that likely was tried to be solved before

nateberkopec commented on lib/puma/thread_pool.rb:
```diff
@@ -208,6 +213,24 @@ def wait_until_not_full
+
+        # this will be signaled once a request finishes,
+        # which can happen earlier than delay
+        @not_full.wait @mutex, BUSY_WORKER_DELAY
```
Are you running this particular variation in prod? It looks like you decided the ticks concept/more-loaded-workers-wait-longer idea was not necessary.

nateberkopec replied:
It would be cool if we could set/get watcher priority in libev through nio4r, then we could do this w/o any latency at all. But that's `TODO`.

ayufan replied:
Not this particular one, once we agree that this is OK. I will push this change to `gitlab.com` and measure it there.

nateberkopec commented:
I need to actually run the bm myself and then check out the `oob` interaction, but overall the change seems very simple and makes sense to me.

ayufan commented:
I updated specs for OOB to have less likely of hitting race condition in tests. Adding small delay seems to do the trick and allows us to properly test all scenarios, by ensuring that requests processing is not "so immediate", as currently `/` for OOB was immediate.

@wjordan @nateberkopec WDYT?

dentarg commented on lib/puma/dsl.rb:
```diff
@@ -664,6 +664,14 @@ def shutdown_debug(val=true)
+    # accepting new socket
+    # if we are already processing requests
+    # this allows less busy workers to pick before us
+    def wait_for_less_busy_worker(val=true)
```
Should we mention here this only does something on MRI?

ayufan replied:
Good idea.

dentarg commented on test/helper.rb:
```diff
@@ -116,6 +116,7 @@ def skip_unless(eng, bt: caller)
       when :darwin  then "Skip unless darwin"  unless RUBY_PLATFORM[/darwin/]
       when :jruby   then "Skip unless JRuby"   unless Puma.jruby?
       when :windows then "Skip unless Windows" unless Puma.windows?
+      when :mri     then "Skip unless MRI" unless Puma.mri?
```
```suggestion
      when :mri     then "Skip unless MRI"     unless Puma.mri?
```

ayufan replied:
Sorry, yes, my omission. Fixing it :)

ayufan commented:
@nateberkopec I asked our production team to change params of old patch: https://gitlab.com/gitlab-org/gitlab/-/issues/196002#note_330784539. The old patch is still equally the same, as we can control all aspects of it, with configuring params. This PR is simplified version with these values hardcoded.

ayufan commented:
Some results from running the benchmark (included): https://docs.google.com/spreadsheets/d/1y2YqrPPgZ-RtjKiCGplJ7YkwjMXkZx6vEPq7prFbUn4/edit#gid=1510215191. The exact numbers are fully dependent on a workload, this hopefully somehow shows that interleaved workflow.

ayufan commented:
@nateberkopec WDYT? What would be the next steps? :)

nateberkopec commented:
I'm gonna take some time this weekend to review 👍 The code change is simple, so the main thing is figuring out the impact and benchmarking.

schneems commented:
> figuring out the impact and benchmarking

Agreed, that's a tall order. We could publish this in a pre-release and then solicit people to try it out, though finding the people and getting meaningful feedback there has always been an issue.  We can promote via twitter and slack etc. it's better than nothing.

nateberkopec commented:
Most def. Once the 5.0 milestone is complete I'll release 5.0.0.beta1

ayufan commented:
@nateberkopec

I rebased on `master` as there were some changes merge related to OOB:

- I added a config setting where you specify delay, which is `off` by default
- This makes it not yet default, but allows us to merge/test without the impact on existing users without them doing a very simple addition
- I reworked a `cpu_spin.rb` to make it easier
- I added additional `test.rb` to very simply test that, since this is `sleep-based` I had to use small `sleeps`, this test is pretty fast anyway: `~0.238589s`

Do you think this to be acceptable in that form to merge and continue testing this non-blocking/non-breaking change?

nateberkopec commented on lib/puma/thread_pool.rb:
```diff
@@ -228,6 +228,25 @@ def wait_until_not_full
+      # Ruby MRI does GVL, this can result
+      # in processing contention when multiple threads
+      # (requests) are running concurrently
+      return unless Puma.mri?
```
This is adding two method calls to the hot path, but for now I'll leave them in. Just pointing out for future optimiation that we could metaprogram these out based on config.

ghiculescu replied:
It might be worth swapping the order of calls.

```
require "benchmark/ips"

def mri?
  RUBY_ENGINE == 'ruby' || RUBY_ENGINE.nil?
end

num = rand(-10..10)

Benchmark.ips do |b|
  b.report("mri?") { mri? }

  b.report("delay_s > 0") { num > 0 }

  b.compare!
end
```

```
Warming up --------------------------------------
                mri?   714.254k i/100ms
         delay_s > 0     1.675M i/100ms
Calculating -------------------------------------
                mri?      7.118M (± 3.3%) i/s -     35.713M in   5.023186s
         delay_s > 0     15.494M (±10.6%) i/s -     77.054M in   5.055471s

Comparison:
         delay_s > 0: 15493847.0 i/s
                mri?:  7118414.2 i/s - 2.18x  (± 0.00) slower
```

(I'm not sure how hot this path runs / how much it matters.)

ghiculescu replied:
https://github.com/puma/puma/pull/2579

matthiassb replied:
Would ruby 2.2.1 be impacted by MRI GVL as well?

nateberkopec commented:
Excellent work @ayufan. I'm going to write up a guide to 5.0 and the beta which will provide more guidance for using this feature, but excited to see reports/data back when this goes live.

ayufan commented:
Thank you @nateberkopec !

mwpastore commented:
Amazing! Thank you all!

ghiculescu commented on lib/puma/server.rb:
```diff
@@ -283,6 +283,9 @@ def handle_servers
                 begin
                   pool.wait_until_not_full
+                  pool.wait_for_less_busy_worker(
+                    @options[:wait_for_less_busy_worker].to_f)
```
There's a micro-optimisation here, which I think is worth flagging only because this happens in a while loop. `to_f` is already called when the option's value is set, [here](https://github.com/puma/puma/pull/2079/files#diff-122c03b9c96f060a9cc29b5ed3446467e1909b0054b4f5cfd95ee41cfff832e7R676).

```ruby
require "benchmark/ips"

options = {foo: 0.5}

Benchmark.ips do |b|
  b.report("to_f on option") { options[:foo].to_f }

  b.report("option directly") { options[:foo] }

  b.compare!
end
```

```
Warming up --------------------------------------
      to_f on option     1.593M i/100ms
     option directly     2.169M i/100ms
Calculating -------------------------------------
      to_f on option     16.280M (± 1.2%) i/s -     82.842M in   5.089545s
     option directly     20.139M (± 8.5%) i/s -    101.943M in   5.103727s

Comparison:
     option directly: 20138636.0 i/s
      to_f on option: 16279512.8 i/s - 1.24x  (± 0.00) slower
```

schneems commented:
I wrote some more details about this behavior in https://github.com/puma/puma/issues/2078. I've also been writing about this optimization and a lot of people are confused about how exactly it works. I want to give a brief recap to attempt to explain how I see it now:

Puma has an "accept loop" (in default boot configuration) where it will continuously call "accept" on the socket independently of the rest of the server. I originally thought it would call once, and block, yielding like IO until the OS got a request and woke it up. It does not do that. Instead, it calls this logic with a timeout of 0, which returns immediately if there are no queued requests. Since it's in a loop, it calls some other housekeeping logic, and then immediately goes back to calling "accept" with a zero timeout.

That means that by default when there are no requests, all workers are racing to be the first one to call "accept" on the socket after the request has been received.

**Before this change:** All workers called "accept" just as fast whether they were fully overloaded (all threads busy) or if they were sitting idle with no requests.

**After this change:** Workers that are 100% idle now loop as fast as possible to race for the incoming connection. Workers who have even 1 busy thread now have a sleep in their loop. This encourages (but does not guarantee) that a completely idle worker will get any new connections first. This is what is meant by "sleep sort."

With Puma 7, we changed the logic such that, instead of sleeping for a fixed amount, the accept loop sleeps proportionally to the load it's under. I.e. a worker with 4 requests being processed will sleep twice as long as a worker with 2. You can see the most recent logic in my PR https://github.com/puma/puma/pull/3746. 

Fundamentally, this series of "sleep sort" optimizations is about prioritizing workers who are less busy to call "accept" more. There might be second and third order interactions such as between the GIL/GVL and OS scheduler, and things like CPU affinity and CPU cache, etc. However, the main logic is centered on slowing down work ingestion for resources that are busier.
